// @wordpress/components/build-module/border-box-control/border-box-control-split-controls/component.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import BorderBoxControlVisualizer from '../border-box-control-visualizer';
import { BorderControl } from '../../border-control';
import { Grid } from '../../grid';
import { contextConnect } from '../../ui/context';
import { useBorderBoxControlSplitControls } from './hook';

const BorderBoxControlSplitControls = (props, forwardedRef) => {
  const {
    centeredClassName,
    colors,
    disableCustomColors,
    enableAlpha,
    enableStyle,
    onChange,
    popoverClassNames,
    value,
    __experimentalHasMultipleOrigins,
    __experimentalIsRenderedInSidebar,
    ...otherProps
  } = useBorderBoxControlSplitControls(props);
  const sharedBorderControlProps = {
    colors,
    disableCustomColors,
    enableAlpha,
    enableStyle,
    isCompact: true,
    __experimentalHasMultipleOrigins,
    __experimentalIsRenderedInSidebar
  };
  return createElement(Grid, _extends({}, otherProps, {
    ref: forwardedRef,
    gap: 4
  }), createElement(BorderBoxControlVisualizer, {
    value: value
  }), createElement(BorderControl, _extends({
    className: centeredClassName,
    hideLabelFromVision: true,
    label: __('Top border'),
    onChange: newBorder => onChange(newBorder, 'top'),
    popoverContentClassName: popoverClassNames === null || popoverClassNames === void 0 ? void 0 : popoverClassNames.top,
    value: value === null || value === void 0 ? void 0 : value.top
  }, sharedBorderControlProps)), createElement(BorderControl, _extends({
    hideLabelFromVision: true,
    label: __('Left border'),
    onChange: newBorder => onChange(newBorder, 'left'),
    popoverContentClassName: popoverClassNames === null || popoverClassNames === void 0 ? void 0 : popoverClassNames.left,
    value: value === null || value === void 0 ? void 0 : value.left
  }, sharedBorderControlProps)), createElement(BorderControl, _extends({
    hideLabelFromVision: true,
    label: __('Right border'),
    onChange: newBorder => onChange(newBorder, 'right'),
    popoverContentClassName: popoverClassNames === null || popoverClassNames === void 0 ? void 0 : popoverClassNames.right,
    value: value === null || value === void 0 ? void 0 : value.right
  }, sharedBorderControlProps)), createElement(BorderControl, _extends({
    className: centeredClassName,
    hideLabelFromVision: true,
    label: __('Bottom border'),
    onChange: newBorder => onChange(newBorder, 'bottom'),
    popoverContentClassName: popoverClassNames === null || popoverClassNames === void 0 ? void 0 : popoverClassNames.bottom,
    value: value === null || value === void 0 ? void 0 : value.bottom
  }, sharedBorderControlProps)));
};

const ConnectedBorderBoxControlSplitControls = contextConnect(BorderBoxControlSplitControls, 'BorderBoxControlSplitControls');
export default ConnectedBorderBoxControlSplitControls;
